### PR TITLE
iOS bug fixes

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/cookies/CookiesStorage.kt [hunk 1]
```diff
@@ -31,7 +31,9 @@ suspend fun CookiesStorage.addCookie(urlString: String, cookie: Cookie) {
 }
 
 internal fun Cookie.matches(requestUrl: Url): Boolean {
-    val domain = domain?.toLowerCasePreservingASCIIRules()?.trimStart('.') ?: error("Domain field should have the default value")
+    val domain = domain?.toLowerCasePreservingASCIIRules()?.trimStart('.')
+        ?: error("Domain field should have the default value")
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/cookies/CookiesStorage.kt]: Could you please improve the message to explain what one should do? Currently, it is not clear what domain we are talking about and where it should be.

File: ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosClientEngine.kt [hunk 2]
```diff
@@ -4,154 +4,74 @@
...
+
+        callContext[Job]!!.invokeOnCompletion {
+            session.finishTasksAndInvalidate()
+        }
+
+        return try {
+            responseReader.awaitResponse()
+        } catch (cause: CancellationException) {
+            if (task.state == NSURLSessionTaskStateRunning) {
+                task.cancel()
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosClientEngine.kt]: Don't we have a race here between `task.resume()` inside of the launching coroutine and cancellation here? Is it legal to invoke `resume()` after `cancel()`?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/cookies/CookiesStorage.kt (hunk 1 above)
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/features/cookies/CookiesStorage.kt]: no problem. btw it doesn't change

File: ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosClientEngine.kt [hunk 3]
```diff
@@ -4,154 +4,74 @@
...
+            }
+            throw cause
+        } catch (cause: IosHttpRequestException) {
+            throw when (cause.origin.code) {
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosClientEngine.kt]: Does the task cancel itself in this case?

File: ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosUtils.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+internal fun ByteArray.toNSData(): NSData = NSMutableData().apply {
+    if (isEmpty()) return@apply
+    this@toNSData.usePinned {
+        appendBytes(it.addressOf(0), size.convert())
+    }
+}
+
+internal fun NSData.toByteArray(): ByteArray {
+    val data: CPointer<ByteVar> = bytes!!.reinterpret()
+    return ByteArray(length.toInt()) { index -> data[index] }
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosUtils.kt]: Does it work well? Shouldn't we use pin + `memcpy` instead?

File: ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosUtils.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+@KtorExperimentalAPI
+@Suppress("KDocMissingDocumentation")
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosUtils.kt]: Why suppress?

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/BenchmarkTest.kt [hunk 6]
```diff
@@ -23,7 +23,7 @@ private val testData = mutableMapOf(
 )
 
 class BenchmarkTest : ClientLoader() {
-    private val size = 1024
+    private val size = 1 // 1024
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/BenchmarkTest.kt]: ?

Comment (e5l) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/BenchmarkTest.kt]: it's not critical for tests, but native works much faster

File: ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosResponseReader.kt [hunk 7]
```diff
@@ -49,19 +55,31 @@ internal class IosResponseReader(private val callContext: CoroutineContext) : NS
...
-        check(chunks.offer(content)) { "Failed to process the received chunk of size: ${content.size}" }
+        try {
+            chunks.offer(content)
+        } catch (cause: CancellationException) {
+            dataTask.cancel()
+        }
     }
 
     override fun URLSession(session: NSURLSession, task: NSURLSessionTask, didCompleteWithError: NSError?) {
         chunks.close()
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosResponseReader.kt]: cancel if `didCompleteWithError`

**Pull request merged**